internal/fromproto5/client_capabilities.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package fromproto5
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+func ReadDataSourceClientCapabilities(in *tfprotov5.ReadDataSourceClientCapabilities) *datasource.ReadClientCapabilities {
+	if in == nil {
+		return nil
+	}
+
+	resp := &datasource.ReadClientCapabilities{
+		DeferralAllowed: in.DeferralAllowed,
+	}
+
+	return resp
+}
+
+func ReadResourceClientCapabilities(in *tfprotov5.ReadResourceClientCapabilities) *resource.ReadClientCapabilities {
+	if in == nil {
+		return nil
+	}
+
+	resp := &resource.ReadClientCapabilities{
+		DeferralAllowed: in.DeferralAllowed,
+	}
+
+	return resp
+}
+
+func ModifyPlanClientCapabilities(in *tfprotov5.PlanResourceChangeClientCapabilities) *resource.ModifyPlanClientCapabilities {
+	if in == nil {
+		return nil
+	}
+
+	resp := &resource.ModifyPlanClientCapabilities{
+		DeferralAllowed: in.DeferralAllowed,
+	}
+
+	return resp
+}
+
+func ImportStateClientCapabilities(in *tfprotov5.ImportResourceStateClientCapabilities) *resource.ImportStateClientCapabilities {
+	if in == nil {
+		return nil
+	}
+
+	resp := &resource.ImportStateClientCapabilities{
+		DeferralAllowed: in.DeferralAllowed,
+	}
+
+	return resp
+}
```

Comment:
This feels like a good place to apply the defaulting, especially because we can be explicit about it here, rather than implicitly letting Go do the defaulting. For example:

```go
func ReadDataSourceClientCapabilities(in *tfprotov5.ReadDataSourceClientCapabilities) datasource.ReadClientCapabilities {
	if in == nil {
		// Client did not indicate any supported capabilities
		return datasource.ReadClientCapabilities{
				DeferralAllowed: false,
		}
	}

	return datasource.ReadClientCapabilities{
		DeferralAllowed: in.DeferralAllowed,
	}
}
```

Then the rest of the Framework logic and structs don't need to use pointers or worry about doing nil checks. (similar thoughts for `fromproto6`)

Reply:
That's a good point, I was just thinking about this since I moved the defaulting to `ConfigureProviderClientCapabilities()` in the automatic provider deferred action implementation since `server_configureprovider.go` doesn't have an equivalent `*fwserver.ConfigureProviderRequest` to return a pointer to.